iceoryx_posh/include/iceoryx_posh/popo/trigger_state.hpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_POSH_POPO_TRIGGER_STATE_HPP
+#define IOX_POSH_POPO_TRIGGER_STATE_HPP
+
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+#include "iceoryx_utils/cxx/function_ref.hpp"
+
+#include <cstdint>
+#include <limits>
+
+namespace iox
+{
+namespace popo
+{
+/// @brief TriggerState holds the state of a trigger and is the base class
+///        for trigger.
+class TriggerState
+{
+  public:
+    static constexpr uint64_t INVALID_TRIGGER_ID = std::numeric_limits<uint64_t>::max();
+    template <typename T>
+    using Callback = void (*)(T* const);
+
+    /// @brief constructs an empty TriggerState
+    TriggerState() = default;
+    virtual ~TriggerState() = default;
+
+    /// @brief constructs a TriggerState object
+    /// @param[in] origin pointer to the origin of the TriggerState
+    /// @param[in] triggerId id of the trigger
+    /// @param[in] callback the callback of the trigger
+    template <typename T>
+    TriggerState(T* const origin, const uint64_t triggerId, const Callback<T> callback) noexcept;
+
+    /// @brief returns the trigger id
+    /// @return the empty TriggerState always returns INVALID_TRIGGER_ID, otherwise the actual triggerId is returned
+    /// which can also be INVALID_TRIGGER_ID
+    uint64_t getTriggerId() const noexcept;
+
+    /// @brief confirms the origin
+    /// @param[in] origin the possible origin
+    /// @return true if the address is equal to the origin, otherwise false. The empty TriggerState returns always
+    /// false.
+    template <typename T>
+    bool doesOriginateFrom(T* const origin) const noexcept;
+
+    /// @brief returns the pointer to the origin.
+    /// @return If T equals the origin type and the TriggerState is not empty it returns the pointer to the origin.
+    /// Otherwise it returns nullptr. If the type is wrong it prints an additional FATAL error message.
```

Comment:
Some weeks ahgo we discussed to remove FATAL and ERROR log messages as normally the error handler should be called when they appear...

Reply:
@budrus You are right but when I am think about this it is more like a common use case. It is like a `dynamic_cast<type*>` where you would like to cast to a type to which it is not castable. In this case `dynamic_cast` returns `nullptr`.

In the end we do exactly the same here. Either you requested the correct type and you get the correct pointer or you request the wrong type and get a nullptr.
The common use case would be something like this:
```cpp
auto subscriber = myTrigger.getOrigin<Subscriber>();
if ( subscriber ) {
   // receive sample
} else {
   // not a subscriber, not interesting for us
}
```


If you agree with this assessment I would remove the `LogFatal()` and update the docu.

Reply:
If the user does soemthing that makes no sense, I would call the error handler. Like we do when the subscriber returns a ptr they never got. Here the API is used wrong and I would prefer to fail fast and hard

Reply:
Error Handler call is added which signals a moderate error. Moderate since the TriggerState can continue working normally and no internal inconsistency are occurring.